Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,261 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
+  · simp only [Fin.mk_one, Fin.isValue, Fin.zero_eta, h.1, Finset.mem_Icc, Left.neg_nonpos_iff,
+    abs_nonneg, and_self]
+  · simp only [Fin.mk_one, Fin.isValue, ← A_c_eq_zero m A h.1, abs_mul, Finset.mem_Icc]
+    constructor
+    · rw [neg_le]
+      apply le_trans (neg_le_abs (A.1 1 1)) ((le_mul_iff_one_le_left h1).mpr h2)
+    · exact le_trans (le_abs_self (A.1 1 1)) ((le_mul_iff_one_le_left h1).mpr h2)
+
+lemma reps_zero_empty : (reps 0) = ∅ := by
+  rw [reps]
+  ext A
+  constructor
+  · intro h
+    have := A_c_eq_zero 0 _ h.1
+    simp only [Fin.isValue, Set.mem_setOf_eq, mul_eq_zero, Set.mem_empty_iff_false] at *
+    rcases this with hl | hr
+    · exfalso
+      simp only [hl, Fin.isValue, lt_self_iff_false, false_and, and_false] at h
+    · simp only [hr, Fin.isValue, abs_zero] at h
+      exact Lean.Omega.Int.le_lt_asymm (abs_nonneg (A.1 0 1)) (h.2.2.2)
+  · simp only [Set.mem_empty_iff_false, Fin.isValue, Set.mem_setOf_eq, false_implies]
+
+noncomputable instance reps_fintype (k : ℤ) : Fintype (reps k) := by
+  by_cases hk : k ≠ 0
+  · let H := Finset.Icc (-|k|) |k|
+    let H4 :=  H × H × H × H
+    apply Fintype.ofInjective (β := H4) (f := fun (M : reps k) =>
+      ⟨⟨M.1.1 0 0,  reps_entries_le_m' k hk M M.2 0 0⟩,
+        ⟨M.1.1 0 1, reps_entries_le_m' k hk M M.2 0 1⟩,
+          ⟨M.1.1 1 0, reps_entries_le_m' k hk M M.2 1 0⟩,
+            ⟨M.1.1 1 1, reps_entries_le_m' k hk M M.2 1 1⟩⟩)
+    intro M N h
+    ext i j
+    rw [Prod.ext_iff] at h
+    simp only [Fin.isValue, Subtype.mk.injEq, Prod.mk.injEq] at h
+    fin_cases i <;> fin_cases j --is there an easy way to golf this?
+    · exact h.1
+    · exact h.2.1
+    · exact h.2.2.1
+    · exact h.2.2.2
+  · simp only [ne_eq, Decidable.not_not] at hk
+    rw [hk, reps_zero_empty]
+    exact Set.fintypeEmpty
+
+lemma reduce_mem_reps (m : ℤ) (hm : m ≠ 0) : ∀ A : Δ m, reduce m A ∈ reps m := by
+  apply reduce_rec
+  · intro A h
+    by_cases h1 : 0 < A.1 0 0
+    · simp only [reps, Fin.isValue, reduce_eqn1 m A h h1, zpow_neg, Set.mem_setOf_eq, smul_coe,
+      coe_inv, coe_T_zpow, adjugate_fin_two_of, neg_zero, cons_mul, Nat.succ_eq_add_one,
+      Nat.reduceAdd, empty_mul, Equiv.symm_apply_apply, of_apply, cons_val', vecMul,
+      cons_dotProduct, vecHead, one_mul, vecTail, Function.comp_apply, Fin.succ_zero_eq_one,
+      neg_mul, dotProduct_empty, add_zero, zero_mul, zero_add, empty_val', cons_val_fin_one,
+      cons_val_one, cons_val_zero, lt_add_neg_iff_add_lt, le_add_neg_iff_add_le]
+      refine ⟨abs_eq_zero.mp h, ?_, ?_, ?_⟩
+      · rw [abs_eq_zero.mp h]; simp only [Fin.isValue, mul_zero, h1]
+      · apply Int.ediv_mul_le; apply A_d_ne_zero m A (by simpa only [abs_eq_zero] using h) hm
+      · rw [mul_comm, ← @Int.sub_eq_add_neg, (Int.emod_def (A.1 0 1) (A.1 1 1)).symm]
+        apply le_trans _ (Int.emod_lt (A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))
+        rw [abs_eq_self.mpr (Int.emod_nonneg (A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))]
```

Comment:
```suggestion
      · rw [mul_comm, ← Int.sub_eq_add_neg, ← Int.emod_def]
        apply le_trans _ (Int.emod_lt (A.1 0 1) (A_d_ne_zero m A (abs_eq_zero.mp h) hm))
        rw [abs_eq_self.mpr (Int.emod_nonneg (A.1 0 1) (A_d_ne_zero m A (abs_eq_zero.mp h) hm))]
```

Reply:
I think that having those extra versions might be overkill, so instead I just left the version thats used the most (the one with `abs`).